Rewrite of the Getting Started guide (#162)
projectsyn/documentation: docs/modules/ROOT/pages/tutorials/getting-started.adoc

= Getting Started with Project Syn

This guide provides an introduction to Project Syn and its various tools. It targets DevOps engineers in charge of large numbers of Kubernetes clusters.

== Why Project Syn?

You are a DevOps engineer in charge of many Kubernetes clusters. You must manually keep them up-to-date, deploy applications in some or all of them, schedule backups, and ensure that all security policies are applied and working. You're overworked and need urgent help.

To make your job easier, you would need:

. A "single point of truth" for your clusters, acting like a "cluster database," including information about each managed cluster: its location, installed applications, settings, and secrets.
. An "agent" in each cluster, proactively listening for changes in the central catalog and performing modifications as required.

> . Finally, a tool to modify the cluster database's status, updating the clusters' records and associated settings.

Reviewer: ```suggestion
. Finally, a tool to modify the clusters' configuartion, updating the clusters' catalog (the Kubernetes objects which get applied to the cluster).
```

Project Syn provides the tools to help you manage many Kubernetes clusters from a central location:

> . Lieutenant keeps the inventory of tenants and clusters, using a set of Git repositories as the storage mechanism;

Reviewer: ```suggestion
. Lieutenant keeps the inventory of tenants and clusters, using a set of Git repositories as the storage for configurations and catalogs;
```

> . Steward acts as the cluster agent, wrapping https://argoproj.github.io/cd/[Argo CD], who keeps an eye on said Git repositories;

Reviewer: ```suggestion
. Steward acts as the cluster agent, wrapping https://argoproj.github.io/cd/[Argo CD], who keeps an eye on the clusters' caltalog Git repository;
```

> . And Commodore serves as the tool used to manipulate the cluster records, saving changes to the Git repositories whenever needed.

Reviewer: ```suggestion
. And Commodore serves as the tool used to update a cluster catalog from its configuration.
```

In this guide, we're going to focus on Lieutenant and Steward. The xref:tutorials/getting-started-2.adoc[second part] will focus on Commodore.

== Requirements

Before you start, please check the following requirements:

> * A GitLab account.

Reviewer: Any specific versions required?

Author: Not that I'm aware of. It works with GitLab.com, which is GitLab 15 afaik.

Reviewer: I was referring to the whole list, not only GitLab. It was just a thought. If versions are not important, leave it as is.

Create the following environment variables. The `GITLAB_TOKEN` variable contains the value of the token created previously.

[source,bash]
--
export GITLAB_USERNAME=johndoe        # Use your own GitLab username
export GITLAB_ENDPOINT=gitlab.com     # Or your own GitLab instance URL (without the "https://" prefix)
export GITLAB_TOKEN=xxxxxxxxxxxxxxxx
--

== Lieutenant

Lieutenant is the central inventory service for all Project Syn-managed clusters. It consists of the https://syn.tools/lieutenant-api/[Lieutenant API] and the https://syn.tools/lieutenant-operator/[Lieutenant Operator]. Lieutenant needs API access to a GitLab server to create and manage Git repositories for clusters and tenants. Lieutenant usually runs in a dedicated Kubernetes cluster, used only to expose its API.

Lieutenant is the central API for all other Project Syn tools. Therefore we will install it first so that all other components can do their job.

Install Lieutenant on the "k3d-lieutenant" cluster with the following command:

[source,bash]
----
curl -fsL https://try.syn.tools/lieutenant | bash
----

> At the end of this procedure, you will have a K3s cluster named "k3d-lieutenant" in your computer, running the Lieutenant API and Operator.

Reviewer: ```suggestion
At the end of this procedure, you will have a K3s cluster named "k3d-lieutenant" on your computer, running the Lieutenant API and Operator.
```

Not sure about that one.

You can explore Lieutenant using `kubectl get tenants`:

[source,bash]
--
kubectl --context k3d-lieutenant -n lieutenant get tenants
--

The command above produces output similar to the following:

[source]
--
NAME             DISPLAY NAME         AGE
t-aged-pond-24   Project Syn Tenant   13s
--

In the context of Project Syn, a "tenant" is an entity to whom clusters are assigned. Tenants could be customers, departments, teams, or any entity owning a cluster.

Project Syn stores cluster-specific configuration values in that tenant's own configuration Git repository. You should now have a new project in your GitLab account called "Project Syn Tenant."

== Steward

> https://syn.tools/steward/[Steward] is the in-cluster agent of Project Syn. It runs in each cluster associated with a tenant, automatically reacting to changes in the configuration stored in the corresponding Git repositories.

Reviewer: ```suggestion
https://syn.tools/steward/[Steward] is the in-cluster agent of Project Syn. It runs in each cluster associated with a tenant, automatically reacting to changes in the catalog Git repository belonging to that cluster.
```

Install Steward in K3s with one simple command:

[source,bash]
--
curl -fsL https://try.syn.tools/steward | bash
--

> The installation of Steward happens on the K3s cluster "k3d-steward" via an install URL provided by Lieutenant, using a one-time bootstrap token. The token is valid only once and for 30 minutes after cluster registration.

Reviewer 2: The bootstrap tokens are valid for up to 24h using recent Lieutenant versions:

```suggestion
The installation of Steward happens on the K3s cluster "k3d-steward" via an install URL provided by Lieutenant, using a one-time bootstrap token. The token is valid only once and for up to 24 hours after cluster registration.
```

[source,bash]
----
kubectl --context k3d-steward -n syn get secret steward -o json | jq -r .data.token | base64 --decode
----

After these steps, the local K3s cluster is now Syn enabled, and a Git repository called "Project Syn Cluster" is available in GitLab. It's worth pointing out that Argo CD hasn't yet synchronized any changes, and the GitLab repository is empty.

== What's Next?

This guide provided a quick overview of Lieutenant and Steward. If you are interested in knowing more about Project Syn, proceed to the xref:tutorials/getting-started-2.adoc[second part] to learn about Commodore, the third major component of Project Syn.

Please don't execute the steps in the following "Cleaning Up" section if you intend to perform the actions of the xref:tutorials/getting-started-2.adoc[second part] of this guide. We will reuse the clusters in their current state.

== Cleaning Up

If you don't wish to proceed to the xref:tutorials/getting-started-2.adoc[second part] of this tutorial, you can clean up all generated clusters and GitLab projects using the following steps:

> . Delete the cluster and tenant objects, removing the GitLab projects at the same time:

Reviewer: Might be worth to add a Danger admonition, warning yet again to only execute if no intention to do the second part.

Author: It's indicated above.

Reviewer: Yes, I know. But you know how things are more than 240 characters… That is too much to remember such an important thing 😛.

= Getting Started with Project Syn

This guide provides an introduction to Project Syn and its various tools. It targets DevOps engineers in charge of large numbers of Kubernetes clusters.

== Why Project Syn?

You are a DevOps engineer in charge of many Kubernetes clusters. You must manually keep them up-to-date, deploy applications in some or all of them, schedule backups, and ensure that all security policies are applied and working. You're overworked and need urgent help.

To make your job easier, you would need:

. A "single point of truth" for your clusters, acting like a "cluster database," including information about each managed cluster: its location, installed applications, settings, and secrets.
. An "agent" in each cluster, proactively listening for changes in the central catalog and performing modifications as required.
. Finally, a tool to modify the cluster database's status, updating the clusters' records and associated settings.

> Project Syn provides the tools to help you manage many Kubernetes clusters from a central location:

Reviewer: This is probably a good place to get visual. Some overview like the one shown on https://syn.tools/syn/about/architecture.html (potentially simplified) helps to get the big picture.

Author: Too much information. I'd rather not.

> . The https://syn.tools/lieutenant-api/[Lieutenant API] and https://syn.tools/lieutenant-operator/[Operator] as the single point of truth;

Reviewer: Either simplify by just speaking of Lieutenant, or explain that the API and the Operator build up a single concept and are separate things only for technical reasons.

Author: I separated them so I could link to both, but I can send that to the bottom.

Author: Sent links to the bottom, simplified this section.

. https://syn.tools/steward/[Steward] as the cluster agent;

> . And https://syn.tools/commodore/[Commodore] as the tool to manipulate the cluster records.

Reviewer 2: I like this way of explaining the components of Project Syn, but it's slightly misleading imho. You don't mention the pile of Git repos which are part of the "single point of truth", giving the impression that all the cluster config is stored in the Lieutenant API / Operator, and then say that Commodore manipulates cluster records, when in reality it primarily works with the Git repos, and not the Lieutenant API.

Author: Good point, will rephrase

Author: Done

> In this guide, we're going to focus on Lieutenant and Steward. The second part will focus on Commodore.

Reviewer: ```suggestion
In this guide, we're going to focus on Lieutenant and Steward. The xref:tutorials/gettings-started-2.adoc[second part] will focus on Commodore.
```

== Requirements

Before you start, please check the following requirements:

* A GitLab account.
* The following tools:
** `kubectl`
** `k3d`
** `jq`
** `curl`

Modify your `/etc/hosts` file and add the following entry:

[source]
--

> 127.0.0.1       host.k3d.internal

Reviewer: Can this work without? For instance, by using the plain IP or xip.io (127.0.0.1.xip.io}?

Author: Nope, sorry. K3s nodes internally use `host.k3d.internal` to talk to the host, and we need to expose locally Lieutenant API without using ngrok or any other reverse proxy (so as to reduce requirements)… if anyone finds a better solution, I'm all ears, but this is the only one that worked for me so far.

Author: (To be more precise: the issue is to access Lieutenant with the same URL both inside and outside of the cluster, that is.)

--

> Launch two new K3s clusters with `k3d`:

Reviewer: Please explain what those clusters are supposed to represent. Managed cluster and cluster we use to manage the fleet.

Author: No, at this point we just ask for requirements, we don't explain why we need them. That comes later.

[source,bash]
--
k3d cluster create lieutenant --port "35777:8080@loadbalancer" --image=rancher/k3s:v1.23.8-k3s1
--

and

[source,bash]
--
k3d cluster create steward  --port "35778:8080@loadbalancer" --image=rancher/k3s:v1.23.8-k3s1
--

In your GitLab account, create a token with the 'API' scope at https://gitlab.com/-/profile/personal_access_tokens.

Create the following environment variables. The `GITLAB_TOKEN` variable contains the value of the token created previously.

[source,bash]
--
export GITLAB_USERNAME=johndoe        # Use your own GitLab username
export GITLAB_ENDPOINT=gitlab.com     # Or your own GitLab instance URL
export GITLAB_TOKEN=xxxxxxxxxxxxxxxx
--

== Lieutenant

Lieutenant is the central inventory service for all Project Syn-managed clusters. It consists of the Lieutenant _Operator_ and the Lieutenant _API_. Lieutenant needs API access to a GitLab server to create and manage Git repositories for clusters and tenants.

> Lieutenant usually runs in a dedicated Kubernetes cluster, used only to expose its API and allow for manipulating Project Syn-managed clusters from a central location.

Reviewer: I feel like the role of Lieutenant is miss presented in this tutorial. By my understanding, Lieutenant is an inventory of managed lusters. I do not use it to manipulate clusters. This I do through the config hierarchy. And that lives in git repositories.

Author: You're right, the wording is not correct.

Author: Corrected

Lieutenant is the central API for all other Project Syn tools. Therefore we will install it first so that all other components can do their job.

Install Lieutenant on the "lieutenant" cluster with the following command:

[source,bash]
----

> curl -fsL https://try.syn.tools/lieutenant | bash

Reviewer 2: Where do the scripts live?

Also this script prints the gitlab token in plaintext in the console, which is maybe not the best idea ever.

Author: Corrected

----

At the end of this procedure, you will have a K3s cluster in your computer, running the Lieutenant API and Operator.

You can explore Lieutenant using `kubectl get tenants` running the following command:

[source,bash]
--
kubectl --context k3d-lieutenant get tenants -n lieutenant
--

The command above returns something like the following:

[source]
--
NAME             DISPLAY NAME         AGE
t-aged-pond-24   Project Syn Tenant   13s
--

> In the context of Project Syn, a "tenant" is an entity to whom clusters are assigned. Tenants could be customers, departments, teams, or any entity responsible for owning a cluster.

Reviewer: ```suggestion
In the context of Project Syn, a "tenant" is an entity to whom clusters are assigned. Tenants could be customers, departments, teams, or any entity owning a cluster.
```

Owning could also be defined as "paying for". 🤷🏼

> Project Syn stores cluster-specific configuration values in that tenant's own configuration Git repository. You should now have a new project in your GitLab account called "Project Syn Tenant."

Reviewer: Was there already an introduction to the hierarchy? I find it important to understand, that the configuration is a hierarchical aggregation of several sources.

Author: What difference does it make that the information is flat or hierarchical? At this point, we should only care that it exists. Again, this is a getting started guide, so less is more.

Reviewer: Valid argument. I do not have a strong opinion here. It was just the first thought I had.

Author: Perfect then, resolving this item

== Steward

> Steward is the in-cluster agent of Project Syn. It runs in each cluster associated with a tenant, automatically reacting to changes in the configuration stored in the corresponding Git repositories.

Reviewer: That is only part of it. The heavy lifting is actually done by Argo CD. Steward only does the initial bootstrap of Argo CD and takes care of facts reporting to Lieutenant.

Author: That's explained below. I'd rather not go into much depth at this level.

Install Steward in K3s with one simple command:

[source,bash]
--

> curl -fsL https://try.syn.tools/steward | bash

Reviewer 2: Waiting for 30 seconds for the bootstrap token to become valid seems excessive, generally, the operator shouldn't take more than 5 seconds to initialize a cluster.

Author: Will reduce the time there then, let's try!

Author: Reduced to 5 seconds, deploying

argocd-redis-6d65dbd578-d2znt         1/1     Running   0          59s
argocd-server-b5f6b5cff-vqq5x         1/1     Running   0          59s
argocd-repo-server-569b975697-647r4   1/1     Running   0          59s
argocd-application-controller-0       1/1     Running   0          57s
--

At this point, Lieutenant is aware of the new K3s cluster, and we can see that through the `kubectl get clusters` command:

[source,bash]
--
kubectl --context k3d-lieutenant get clusters -n lieutenant
--

The command above outputs something similar to this:

[source]
--
NAME                  DISPLAY NAME          TENANT                  AGE
c-long-firefly-9017   Project Syn Cluster   t-autumn-silence-6912   12s
--

Cluster names have the `c-` prefix, while tenant names have with `t-` prefix.

=== Accessing Argo CD

As you have seen above, Steward wraps https://argoproj.github.io/cd/[Argo CD], a Cloud-Native continuous deployment and integration tool, continuously observing the GitLab repositories for changes.

You can also connect to Argo CD and see its console. Start by exposing the `argocd-server` deployment:

[source,bash]
----

> kubectl --context k3d-steward -n syn expose deployment argocd-server --type=LoadBalancer --port=8080

Reviewer 2: We should create a separate LB service here, since the `argocd-server` service will become managed by ArgoCD in part 2.

```suggestion
kubectl --context k3d-steward -n syn expose deployment argocd-server --type=LoadBalancer --port=8080 --name=argocd-server-lb --overrides='{"metadata":{"labels":{"argocd.argoproj.io/instance":null}}}'
```

The `--overrides` bit is necessary to remove the `argocd.argoproj.io/instance` label from the Service object (which the service inherits from the deployment through `kubectl expose`), so ArgoCD doesn't remove it.

----

Open Argo CD in your browser at http://localhost:35778. Log in to Argo CD with the username `admin` and the password returned by this command:

[source,bash]
----
kubectl --context k3d-steward -n syn get secret steward -o json | jq -r .data.token | base64 --decode

> ----

Reviewer 2: Might be worth to note here that it's expected that ArgoCD isn't synced at this point.

Author: Done

> After these steps, the local K3s cluster is now Syn enabled, and a Git repository called "Project Syn Cluster" is stored in GitLab.

Reviewer 2: ```suggestion
After these steps, the local K3s cluster is now Syn enabled, and a Git repository called "Project Syn Cluster" is available in GitLab.
```
